go/ql/src/experimental/CWE-321/HardcodedKeysLib.qll

```diff
@@ -0,0 +1,291 @@
+      |
+        this = m.getArgument(1)
+      )
+    }
+  }
+
+  private class GoKitJwt extends Sink {
+    GoKitJwt() {
+      exists(DataFlow::CallNode m |
+        // `func NewSigner(kid string, key []byte, method jwt.SigningMethod, claims jwt.Claims) endpoint.Middleware`
+        m.getTarget().hasQualifiedName("github.com/go-kit/kit/auth/jwt", "NewSigner")
+      |
+        this = m.getArgument(1)
+      )
+    }
+  }
+
+  private class LestrratJwk extends Sink {
+    LestrratJwk() {
+      exists(DataFlow::CallNode m, string pkg |
+        pkg.matches([
+            "github.com/lestrrat-go/jwx", "github.com/lestrrat/go-jwx/jwk",
+            "github.com/lestrrat-go/jwx%/jwk"
+          ]) and
+        // `func New(key interface{}) (Key, error)`
+        m.getTarget().hasQualifiedName(pkg, "New")
+      |
+        this = m.getArgument(0)
+      )
+    }
+  }
+
+  /**
+   * Mark any comparision expression where any operand is tainted as a
+   * sanitizer for all instances of the taint
+   */
+  private class CompareExprSanitizer extends Sanitizer {
+    CompareExprSanitizer() {
+      exists(BinaryExpr c |
+        c.getAnOperand().getGlobalValueNumber() = this.asExpr().getGlobalValueNumber()
+      )
+    }
+  }
+
+  /** Mark an empty string returned with an error as a sanitizer */
+  private class EmptyErrorSanitizer extends Sanitizer {
+    EmptyErrorSanitizer() {
+      exists(ReturnStmt r, DataFlow::CallNode c |
+        c.getTarget().hasQualifiedName("errors", "New") and
+        r.getNumChild() > 1 and
+        r.getAChild() = c.getAResult().getASuccessor*().asExpr() and
+        r.getAChild() = this.asExpr()
+      )
+    }
+  }
+
+  /** Mark any formatting string call as a sanitizer */
+  private class FormattingSanitizer extends Sanitizer {
+    FormattingSanitizer() { exists(Formatting::StringFormatCall s | s.getAResult() = this) }
+  }
+
+  /**
+   * Mark any taint arising from a read on a tainted slice with a random index as a
+   * sanitizer for all instances of the taint
+   */
+  private class RandSliceSanitizer extends Sanitizer {
+    RandSliceSanitizer() {
+      // Sanitize flows like this:
+      // func GenerateCryptoString(n int) (string, error) {
+      //   const chars = "123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-"
+      //   ret := make([]byte, n)
+      //   for i := range ret {
+      //     num, err := crand.Int(crand.Reader, big.NewInt(int64(len(chars))))
+      //     if err != nil {
+      //       return "", err
+      //     }
+      //     ret[i] = chars[num.Int64()]
+      //   }
+      //   return string(ret), nil
+      // }
+      exists(
+        DataFlow::CallNode randint, string name, DataFlow::ElementReadNode r, DataFlow::Node index
+      |
+        (
+          randint.getTarget().hasQualifiedName("math/rand", name) or
+          randint.getTarget().(Method).hasQualifiedName("math/rand", "Rand", name) or
+          randint.getTarget().hasQualifiedName("crypto/rand", "Int")
```

Comment:
Nitpick: since `name` isn't used in this disjunct, it's better to pull out the list of names as `getARandIntFunctionName()` and use it in the two disjuncts that need it. This may stop the evaluator from pointlessly forming the cartesian product of `randint.getTarget().hasQualifiedName("crypto/rand", "Int")` and `name = [ ... ]`, though in this particular instance it wouldn't hurt too much since it would only mean 12 duplicates of the set of random int function calls.